### PR TITLE
Slight better language system + New language "Metalanguage" + animals say in metalanguage

### PR DESCRIPTION
**Pull request by EvilDragonfiend**
## About The Pull Request
* Slightly improves language codes
* Chnages defines a bit
* Adds metalanguage - everyone can understand this, but nobody will be capable of speaking this.
* a feature - gibbering (aphasia) language icon is no longer visible to yourself

This PR is necessary to make #https://github.com/BeeStation/BeeStation-Hornet/pull/8824 better for revenant telepathy ability. Having a language is necessary, as taking a random language is quite meh to the telepathy code.

## Why It's Good For The Game
Better communication system is good for RP

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

image

image

ghost and admin can see it in the list

--------------

image

image

image

people see this without language icon, but ghosts can see the icon
and you can see it in the language list if you have it (or you are admin)

-----------

image

You can understand metalanguage even if you can't speak (but you know it)

----------

image

Gibbering

</details>

## Changelog
:cl:
code: improved language codes
add: added Metalanguage - everyone can understand this, but only speakable by specific situation. (mentor rat can say this. curator can't say this). You know this language, but it will not be visible in your known language list.
imageadd: lang icon for metalanguage
tweak: Gibbering language icon will not be visible to yourself any longer. No meta yourself that your language is broken.
tweak: some mobs (especially animals) will now say in Metalanguage instead of Common. Rats will squeak, cats will meow, but not in Galactic Common.
/:cl:

### CODE REVIEW

**Comment by Tsar-Salat** [comment]
> Metalanguage

lmao I keep reading this as "Metal Language" 

Looks like a good way to translate telepathy in to runechat though, good job

**Comment by itsmeow** [comment]
Metalanguage should not be visible in the UI at all imo

**Comment by EvilDragonfiend** [comment]
> Metalanguage should not be visible in the UI at all imo

It should be visible when some admin fuckery is applied.

**Comment by EvilDragonfiend** [comment]
> Metalanguage should not be visible in the UI at all imo

Well, I changed this a bit now.
* Metalanguage is only visible when you're admin or possess `TRAIT_METALANGUAGE_KEY_ALLOWED`
* You can't speak Metalanguage through the key ,` in case you don't possess the trait
* You can speak Metalanguage if your default language is set to it
* trait holder (`TRAIT_METALANGUAGE_KEY_ALLOWED`) can freely change their language to metalanguage
* admin will be capable of giving `TRAIT_METALANGUAGE_KEY_ALLOWED` trait to them

**Comment by EvilDragonfiend** [comment]
Metalanguage is somewhat bad to read and say. I think I should have a replacement.

File: code/__DEFINES/language.dm [hunk 1]
```diff
@@ -1,8 +1,9 @@
 #define NO_STUTTER 1
 #define TONGUELESS_SPEECH 2
-#define LANGUAGE_HIDE_ICON_IF_UNDERSTOOD 4
-#define LANGUAGE_HIDE_ICON_IF_NOT_UNDERSTOOD_WITH_LINGUIST_TRAIT 8 // even if you have linguist trait, you shouldn't see the icon
-#define LANGUAGE_ALWAYS_SHOW_ICON_IF_NOT_UNDERSTOOD 16
+#define LANGUAGE_HIDE_ICON_IF_YOU_UNDERSTAND 4
+#define LANGUAGE_HIDE_ICON_IF_YOU_SHOULD_NOT_RECOGNISE_WITH_LINGUIST_TRAIT 8 // even if you have linguist trait, you shouldn't see the icon
+#define LANGUAGE_ALWAYS_SHOW_ICON_IF_YOU_DO_NOT_UNDERSTAND 16
```
Comment (tontyGH) [inline code/__DEFINES/language.dm]: This is kind of an unnecessary change. The defines were already pretty clear before, now they're just longer

The linguist one is particularly painful, please use comments instead
```suggestion
///Language icon will be hidden if you understand it, example: common
#define LANGUAGE_HIDE_ICON_IF_UNDERSTOOD 4
///Language icon will be hidden even if you have the linguist trait, example: agneusia 
#define LANGUAGE_ALWAYS_HIDE_ICON_IF_NOT_UNDERSTOOD 8
///Language icon will always be visible if regardless of if you have the linguist trait, example: common
#define LANGUAGE_ALWAYS_SHOW_ICON_IF_NOT_UNDERSTOOD 16
//etc...
```
Something like this would be better
Now it's reasonably short and if someone doesn't understand how this flag is used they can just hover over it

File: code/__DEFINES/language.dm [hunk 2]
```diff
@@ -1,8 +1,9 @@
...
+#define LANGUAGE_ALWAYS_SHOW_ICON_TO_GHOSTS 32
```
Comment (tontyGH) [inline code/__DEFINES/language.dm]: It would be nice if these were proper bit shifts instead of numbers

File: code/modules/mob/say.dm [hunk 3]
```diff
@@ -168,6 +168,9 @@
 					// No, you cannot speak in xenocommon just because you know the key
 					if(!can_speak_language(LD))
 						return message
+					// you are not allowed to use metalanguage key
+					if(LD == /datum/language/metalanguage && !HAS_TRAIT(src, TRAIT_METALANGUAGE_KEY_ALLOWED))
+						return message
```
Comment (tontyGH) [inline code/modules/mob/say.dm:173]: Should probably move this into the  can_speak_language proc

**Review by tontyGH — CHANGES_REQUESTED** [review]

**Comment by tontyGH** [comment]
Metalanguage seems alright to me, though having it written as meta_language would probably help

File: code/modules/mob/say.dm (hunk 3 above)
Comment (EvilDragonfiend) [reply code/modules/mob/say.dm:173]: It's a different check if you're really allowed to do that so because it shouldn't be visible or controllable at most times.
What I am intending is something like 'You are even not aware of that you're talking in a metalanguage'

File: code/__DEFINES/language.dm (hunk 1 above)
Comment (EvilDragonfiend) [reply code/__DEFINES/language.dm]: alright

**Comment by EvilDragonfiend** [comment]
Commit contents:
* changes DEFINE names again
* Aphasia language icon is no longer visible to yourself

**Comment by EvilDragonfiend** [comment]
>  Metalanguage seems alright to me, though having it written as meta_language would probably be better

`Metalanguage` is an existing word, and technically closer enough to the meaning of the conceptual word.

**Comment by EvilDragonfiend** [comment]
hmm, actually, I should do something instead of trait....

**Comment by EvilDragonfiend** [comment]
MC caused by cortex PR, and solved.

File: code/datums/brain_damage/mrat.dm [hunk 4]
```diff
@@ -122,6 +122,8 @@
 	leave.Grant(src)
 
 	grant_all_languages()
+	grant_language(/datum/language/metalanguage)
```
Comment (itsmeow) [inline code/datums/brain_damage/mrat.dm]: this should be set to the default language too. Either way it's only visible to the person hearing it. No need to grant all languages in this case either.

Imaginary friends should act the same imo

File: code/modules/language/language.dm [hunk 5]
```diff
@@ -27,17 +27,32 @@
 	var/icon = 'icons/misc/language.dmi'
 	var/icon_state = "popcorn"
 
+/// Returns TRUE/FALSE based on seeing a language icon is validated to a given hearer in the parameter.
 /datum/language/proc/display_icon(atom/movable/hearer)
+ 	// ghosts want to know how it is going.
+	if(flags & LANGUAGE_ALWAYS_SHOW_ICON_TO_GHOSTS && \
```
Comment (itsmeow) [inline code/modules/language/language.dm]: add parenthesis around flag checks or use CHECK_BITFIELD()

File: code/modules/language/language_menu.dm [hunk 6]
```diff
@@ -21,6 +21,7 @@
 /datum/language_menu/ui_data(mob/user)
 	var/list/data = list()
 
+	var/is_admin = check_rights_for(user.client, R_ADMIN)
```
Comment (itsmeow) [inline code/modules/language/language_menu.dm]: this should probably include R_DEBUG for seeing it, not granting though since they can't use the language menu for that

File: code/modules/language/language_menu.dm [hunk 7]
```diff
@@ -83,9 +88,19 @@
...
-				language_holder.selected_language = language_datum
-				. = TRUE
+			if(language_datum)
+				// they're changing their language to something else from metalanguage. It must be mistake.
+				if(language_holder.selected_language == /datum/language/metalanguage && \
+						language_datum != /datum/language/metalanguage && \
+						!HAS_TRAIT(user, TRAIT_METALANGUAGE_KEY_ALLOWED) && \
+						!is_admin)
+					var/no = alert(user, "You're giving up your power to speak in a powerful language that everyone understands. Do you really wish to do that?", "WARNING!", "Yes", "No")
+					if(no == "No")
```
Comment (itsmeow) [inline code/modules/language/language_menu.dm]:
```suggestion
					if(no != "Yes")
```
This way is stabler since X causes alert() to return null

File: code/modules/mob/living/carbon/say.dm [hunk 8]
```diff
@@ -19,9 +19,9 @@
...
+	if(CHECK_BITFIELD(initial(dt.flags), TONGUELESS_SPEECH))
+		return TRUE
 	var/obj/item/organ/tongue/T = getorganslot(ORGAN_SLOT_TONGUE)
 	if(T)
-		. = T.could_speak_language(dt)
-	else
-		. = CHECK_BITFIELD(initial(dt.flags), TONGUELESS_SPEECH)
-
+		return T.could_speak_language(dt)
+
```
Comment (itsmeow) [inline code/modules/mob/living/carbon/say.dm]: add an explicit false return for clarity

**Review by itsmeow — CHANGES_REQUESTED** [review]

**Comment by itsmeow** [comment]
I really like how this is implemented, great job. Good tool to have as well.

File: code/datums/brain_damage/mrat.dm (hunk 4 above)
Comment (EvilDragonfiend) [reply code/datums/brain_damage/mrat.dm]: imaginary friends are a kind of their another personality which means they should inherent the language of their 'original'.
mentor rat is fine with it, so I will update.

**Comment by EvilDragonfiend** [comment]
review addressed, but it seems imaginary mob types need some more touch since they talk in no language option... I will do that a later PR.

**Comment by itsmeow** [comment]
Can you make mice Squeak in metalanguage, pets meow etc

File: code/modules/language/language.dm [hunk 9]
```diff
@@ -30,8 +30,8 @@
 /// Returns TRUE/FALSE based on seeing a language icon is validated to a given hearer in the parameter.
 /datum/language/proc/display_icon(atom/movable/hearer)
  	// ghosts want to know how it is going.
-	if(flags & LANGUAGE_ALWAYS_SHOW_ICON_TO_GHOSTS && \
-			(isobserver(hearer)) || HAS_TRAIT(hearer, TRAIT_METALANGUAGE_KEY_ALLOWED))
+	if((flags & LANGUAGE_ALWAYS_SHOW_ICON_TO_GHOSTS) && \
+			(isobserver(hearer) || (HAS_TRAIT(hearer, TRAIT_METALANGUAGE_KEY_ALLOWED) && src.type == /datum/language/metalanguage)))
```
Comment (itsmeow) [inline code/modules/language/language.dm]: use istype()

File: code/modules/language/language_holder.dm [hunk 10]
```diff
@@ -218,19 +218,24 @@ Key procs
 
 /// Copies all languages from the supplied atom/language holder. Source should be overridden when you
 /// do not want the language overwritten by later atom updates or want to avoid blocked languages.
-/datum/language_holder/proc/copy_languages(var/datum/language_holder/from_holder, source_override)
+/datum/language_holder/proc/copy_languages(var/datum/language_holder/from_holder, source_override=FALSE, spoken=TRUE, understood=TRUE, blocked=TRUE)
 	if(source_override)	//No blocked languages here, for now only used by ling absorb.
```
Comment (itsmeow) [inline code/modules/language/language_holder.dm]: this is the same as the blocked argument

**Review by itsmeow — CHANGES_REQUESTED** [review]

File: code/modules/language/language_holder.dm (hunk 10 above)
Comment (EvilDragonfiend) [reply code/modules/language/language_holder.dm]: these two arguments serve different purposes. At least I changed the logic, so you can recognise how it's different.

**Comment by EvilDragonfiend** [comment]
addressed.

image

simple mobs will say in a specified language by a new var `var/speak_language`

**Review by itsmeow — APPROVED** [review]
Based as hell
